### PR TITLE
Add basic xattr APIs

### PR DESCRIPTION
**Pull request by prashanthpai**
Signed-off-by: Prashanth Pai <ppai@redhat.com>

### CODE REVIEW

File: gfapi/fd.go [hunk 1]
```diff
@@ -118,3 +118,58 @@ func (fd *Fd) lseek(offset int64, whence int) (int64, error) {
...
+	var ret C.ssize_t
+	var err error
+
+	attrp, err := syscall.BytePtrFromString(attr)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(dest) <= 0 {
+		ret, err = C.glfs_fgetxattr(fd.fd, (*C.char)(unsafe.Pointer(attrp)), nil, 0)
```
Comment (kshlm) [inline gfapi/fd.go]: The string pointers look so complex. Why don't you just use `C.CString()` to get C compatible string pointers. That's what is being used elsewhere in the package.

**Comment by kshlm** [comment]
Everything other than the string pointers look good to me.

File: gfapi/fd.go (hunk 1 above)
Comment (prashanthpai) [inline gfapi/fd.go]: Thanks, will use `C.CString`. Was referring to conventions followed [here](https://golang.org/src/syscall/syscall.go)

**Comment by prashanthpai** [comment]
On a second thought...

Making use of either `C.CString` or `BytePtrFromString`, both involves making a copy of string data. There is no workaround for that. 

However, the downside of using `C.CString` is that:
* We will have to free memory ourselves using `C.free`
* There is `malloc()` and `free()` overhead at entry and exit points of every go function.

If we use `BytePtrFromString` instead, we do not have to free the memory allocated and there is no immediate `malloc()` and `free()` overhead. As `BytePtrFromString` internally uses `make()` to allocate memory, it will be GC'd at some point later.

**Comment by kshlm** [comment]
C.CString allocates memory for the string the C heap, unlike BytePtrFromString which does it in the Go memory space. This makes the C.CString string safer to be used in C, as the Go runtime is less likely to be affected by misbehaving C code.

**Comment by kshlm** [comment]
Looks good to me. Thanks @prashanthpai !

**Comment by prashanthpai** [comment]
@kshlm Isn't it pretty safe to use it where the underlying C function accepts `const char *` ?  Besides, through out the code base, C strings is the only data structure whose memory is allocated in the C heap. All data buffers in I/O path are in the Go heap.

For example, the `Read` function in `fd.go`, it takes a `[]byte` instance allocated in Go heap. This is directly passed to underlying C code.

I'm not suggesting to change that. Just pointing out that only C strings are in the C heap, and it incurs malloc() + free() overhead, where as everything else in the process runtime resides in the Go heap.

**Pull request merged**